src/squlearn/util/executor.py

```diff
@@ -356,27 +431,238 @@ def __init__(
         if self._caching:
             self._cache = ExecutorCache(self._logger, cache_dir)
 
-        self._logger.info(f"Executor initialized with backend: {{}}".format(self._backend))
-        self._logger.info(f"Executor initialized with service: {{}}".format(self._service))
+        self._logger.info(f"Executor initialized with {{}}".format(self.quantum_framework))
+        if self._backend is not None:
+            self._logger.info(f"Executor initialized with backend: {{}}".format(self._backend))
+        if self._service is not None:
+            self._logger.info(f"Executor initialized with service: {{}}".format(self._service))
         if self._session is not None:
             self._logger.info(
                 f"Executor initialized with session: {{}}".format(self._session.session_id)
             )
-        else:
-            self._logger.info(f"Executor initialized with session: {{}}".format(self._session))
-        self._logger.info(f"Executor initialized with estimator: {{}}".format(self._estimator))
-        self._logger.info(f"Executor initialized with sampler: {{}}".format(self._sampler))
+        if self._estimator is not None:
+            self._logger.info(f"Executor initialized with estimator: {{}}".format(self._estimator))
+        if self._sampler is not None:
+            self._logger.info(f"Executor initialized with sampler: {{}}".format(self._sampler))
         self._logger.info(f"Executor intial shots: {{}}".format(self._inital_num_shots))
 
+    @property
+    def quantum_framework(self) -> str:
+        """Return the quantum framework that is used in the executor.
+
+        Either "qiskit" or "pennylane"
+        """
+        return self._quantum_framework
+
+    @property
+    def pennylane_device(self):
+        """Returns the PennyLane device that is used in the executor."""
+        if self.quantum_framework != "pennylane":
+            raise RuntimeError("PennyLane device is only available for PennyLane backends")
+        return self._pennylane_device
+
+    def pennylane_decorator(self, pennylane_function):
+        """
+        Decorator for a PennyLane circuit
+
+        Args:
+            pennylane_function (callable): The PennyLane circuit function
+        """
+        return qml.qnode(self.pennylane_device, diff_method="best")(pennylane_function)
+
+    def pennylane_execute(self, pennylane_circuit: callable, *args, **kwargs):
```

Comment:
pennylane_run sound good. However, backend_run and pennylane_run have fundermental different interfaces, since pennylane circuits are callables and qiskit circuits is a container. What about pennylane_run and qiskit_run?

Reply:
changed backend_run to qiskit_execute